### DISCUSSION TITLE
How to call eth_signTypedData_v4 in integration test?

### DISCUSSION THREAD

**Discussion by stefan-girlich** [wevm/wagmi#579] [General]
I am trying to invoke different wagmi functions (e.g. `signTypedData`) in a Jest-driven test suite for integration testing. I want to sign some data and finally invoke a smart contract on Rinkeby.

My [test setup](https://pastebin.com/2D3FKkik) is largely derived from `wagmi`'s codebase and just hardcodes Rinkeby as the chain to use. My [test code](https://pastebin.com/2D3FKkik) calls `signTypedData` and is supposed to invoke other `wagmi` functions in the future, including sending transactions.

As of now, both RPC providers (Alchemy and Infura) will tell me that the method `eth_signTypedData_v4` is not available. Example response from Infura:
```
  reason: 'processing response error',
  code: 'SERVER_ERROR',
  body: '{"jsonrpc":"2.0","id":42,"error":{"code":-32601,"message":"The method eth_signTypedData_v4 does not exist/is not available"}}',
  requestMethod: 'POST',
  url: 'https://rinkeby.infura.io/v3/<MY_PROJECT_ID>'
```

1. Would you know why both providers do not provide the signing function via their APIs?
2. My test wallet exists on Rinkeby and I instantiate it via `const wallet = new Wallet({privateKey, address}); return provider.getSigner(wallet.address)` (see linked code). Since [Metamask uses Infura by default](https://metamask.zendesk.com/hc/en-us/articles/4417315392795-What-is-Infura-and-why-does-MetaMask-use-it-), I assume that I am simply unauthorized to call _eth_signTypedData_v4_ via API. How would I provide the right auth information in this case?

Thanks so much for your help!

**Comment by stefan-girlich** [comment]
### Root cause
The test setup I adapted [from the wagmi codebase](https://github.com/tmm/wagmi/blob/main/packages/core/test/utils.ts#L155) calls `provider.getSigner(wallet.address)` to create a `JsonRpcSigner` which will try to invoked `eth_signTypedData_v4` via Infura/Alchemy, both of which don't provide this method in their respective APIs:

```
export function getSigners() {
    const provider = getProvider()
    const signers = accounts.map(x => {
        const wallet = new Wallet({
            privateKey: x.privateKey,
            address: x.address,
        })
        return provider.getSigner(wallet.address)
    })
    return signers
}
```

### Workaround
```
export function getSigners() {
    const signers = accounts.map(x => {
        return new Wallet({
            privateKey: x.privateKey,
            address: x.address,
        })
    })
    return signers
}
```

### Possible aftereffect
I expect on-chain operations (e.g. sending transactions) to fail since the wallet does not feature a provider.
=> Which approach should be used to enable both local signing and on-chain operations?

**Reply by Na3aga** [reply]
@stefan-girlich did you manage to write automated tests with local signing ?

**Reply by 0xPro** [reply]
i'd like to know as well please
